03_ROS/ghost_sim/src/v5_robot_plugin.cpp

```diff
@@ -193,68 +257,153 @@ void V5RobotPlugin::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf){
+	auto motor_data = impl_->actuator_jacobian_.inverse() * joint_data;
+	return motor_data;
+}
+
+// Wraps encoder matrix into V5Encoder state msg[21]
+void V5RobotPlugin::wrapEncoderMsg(){
 	int col_index = 0;
-	// Lamda for-each expression to get encoder values for every joint
-	for_each(begin(impl_->joint_msg_list_), end(impl_->joint_msg_list_), [&](const std::string &joint_data){
-			// Get iterator that points to corresponding joint name in joint_names_
-			auto joint_name_itr = find(impl_->joint_names_.begin(), impl_->joint_names_.end(), joint_data);
 
-			// Convert iterator to index
-			size_t jacobian_row_index = distance(begin(impl_->joint_names_), joint_name_itr);
+	for_each(begin(impl_->encoder_names_), end(impl_->encoder_names_), [&](const std::string &encoder_name){
+			auto encoder_vel_data = impl_->encoder_velocities_;
+			auto encoder_pos_data = impl_->encoder_positions_;
+			impl_->encoder_msg_[col_index].device_name = encoder_name;
+			impl_->encoder_msg_[col_index].device_id = 1;
+			impl_->encoder_msg_[col_index].device_connected = true;
+			impl_->encoder_msg_[col_index].position_degrees = encoder_pos_data(col_index);
+			impl_->encoder_msg_[col_index].velocity_rpm = encoder_vel_data(col_index);
+			impl_->encoder_msg_[col_index].torque_nm = 0.0;
+			impl_->encoder_msg_[col_index].voltage_mv = 0.0;
+			impl_->encoder_msg_[col_index].current_ma = 0.0;
+			impl_->encoder_msg_[col_index].power_w = 0.0;
+			impl_->encoder_msg_[col_index].temp_c = 0.0;
 
-			// impl_->encoder_vel_data_[col_index] = impl_->model_->GetJoint(joint_data)->Position() * impl_->sensor_jacobian_(jacobian_row_index, col_index);
 			col_index++;
-			impl_->encoder_msg_ = this->wrapEncoderMsg(col_index);
 		});
 }
 
-ghost_msgs::msg::V5EncoderState V5RobotPlugin::wrapEncoderMsg(const int col_index){
-	auto msg = ghost_msgs::msg::V5EncoderState();
-	msg.device_name = impl_->encoder_names_[col_index];
-	msg.device_id = 1;
-	msg.device_connected = true;
-	msg.position_degrees = 30.0;
-	// msg.position_degrees = (encoder_vel_data_[col_index] - last_encoder_vel_data_[col_index]) * dt;
-	msg.velocity_rpm = impl_->encoder_vel_data_(col_index);
-	msg.torque_nm = 0.0;
-	msg.voltage_mv = 0.0;
-	msg.current_ma = 0.0;
-	msg.power_w = 0.0;
-	msg.temp_c = 0.0;
-
-	return msg;
-}
+void V5RobotPlugin::populateSensorMsg(){
+	impl_->sensor_msg_.header.stamp = impl_->clock_.now();
+	impl_->sensor_msg_.header.frame_id = "base";
+	impl_->sensor_msg_.msg_id = 1;
+	impl_->sensor_msg_.encoders = impl_->encoder_msg_;
 
-ghost_msgs::msg::V5SensorUpdate V5RobotPlugin::populateSensorMsg(){
-	auto msg = ghost_msgs::msg::V5SensorUpdate();
-	// msg.header =;
-	// msg.msg_id = 1;
-	// msg.encoders =
+	// hardware parameters
+	impl_->sensor_msg_.digital_port_vector[8] = (false, false, false, false, false, false, false, false);
 
-	return msg;
+	impl_->sensor_msg_.joystick_msg = impl_->joystick_msg_;
 }
 
+// Preserves order of joints listed in xacro
 void V5RobotPlugin::getJointStates(){
```

Comment:
could this be updateJointStates(), since it isn't returning anything?

Reply:
Changed!